cdc/kv/shared_client.go

```diff
@@ -0,0 +1,1184 @@
+		}
+		if compatibility := innerErr.GetCompatibility(); compatibility != nil {
+			return cerror.ErrVersionIncompatible.GenWithStackByArgs(compatibility)
+		}
+		if mismatch := innerErr.GetClusterIdMismatch(); mismatch != nil {
+			return cerror.ErrClusterIDMismatch.GenWithStackByArgs(mismatch.Current, mismatch.Request)
+		}
+
+		log.Warn("empty or unknown cdc error",
+			zap.String("namespace", s.changefeed.Namespace),
+			zap.String("changefeed", s.changefeed.ID),
+			zap.Stringer("error", innerErr))
+		metricFeedUnknownErrorCounter.Inc()
+		s.scheduleRegionRequest(ctx, errInfo.singleRegionInfo)
+		return nil
+	case *rpcCtxUnavailableErr:
+		metricFeedRPCCtxUnavailable.Inc()
+		s.scheduleDivideRegionAndRequest(ctx, errInfo.span, errInfo.requestedTable)
+		return nil
+	case *sendRequestToStoreErr:
+		metricStoreSendRequestErr.Inc()
+		bo := tikv.NewBackoffer(ctx, tikvRequestMaxBackoff)
+		s.regionCache.OnSendFail(bo, errInfo.rpcCtx, regionScheduleReload, err)
+		s.scheduleRegionRequest(ctx, errInfo.singleRegionInfo)
+		return nil
+	default:
+		// TODO(qupeng): for some errors it's better to just deregister the region from TiKVs.
+		log.Warn("event feed meets an internal error, fail the changefeed",
+			zap.String("namespace", s.changefeed.Namespace),
+			zap.String("changefeed", s.changefeed.ID),
+			zap.String("span", errInfo.requestedTable.span.String()),
+			zap.Uint64("requestID", errInfo.requestedTable.requestID),
+			zap.Error(err))
+		return err
+	}
+}
+
+func (s *SharedClient) resolveLock(ctx context.Context) error {
+	resolveLastRun := make(map[uint64]time.Time)
+
+	gcResolveLastRun := func() {
+		if len(resolveLastRun) > 1024 {
+			copied := make(map[uint64]time.Time)
+			now := time.Now()
+			for regionID, lastRun := range resolveLastRun {
+				if now.Sub(lastRun) < resolveLockMinInterval {
+					resolveLastRun[regionID] = lastRun
+				}
+			}
+			resolveLastRun = copied
+		}
+	}
+
+	doResolve := func(regionID uint64, state *regionlock.LockedRange, maxVersion uint64) {
+		if state.CheckpointTs.Load() > maxVersion || !state.Initialzied.Load() {
+			return
+		}
+		if lastRun, ok := resolveLastRun[regionID]; ok {
+			if time.Since(lastRun) < resolveLockMinInterval {
+				return
+			}
+		}
+		start := time.Now()
+		defer s.metrics.lockResolveRunDuration.Observe(float64(time.Since(start).Milliseconds()))
+
+		if err := s.lockResolver.Resolve(ctx, regionID, maxVersion); err != nil {
+			log.Warn("event feed resolve lock fail",
+				zap.String("namespace", s.changefeed.Namespace),
+				zap.String("changefeed", s.changefeed.ID),
+				zap.Uint64("regionID", regionID),
+				zap.Error(err))
+		}
+		resolveLastRun[regionID] = time.Now()
+	}
+
+	gcTicker := time.NewTicker(resolveLockMinInterval * 3 / 2)
+	defer gcTicker.Stop()
+LOOP:
+	for {
+		var task resolveLockTask
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-gcTicker.C:
+			gcResolveLastRun()
+			continue LOOP
```

Comment:
There is only one layer of loop, why LOOP is needed?

Reply:
I'm not sure whether `continue` can jump out from `select` block or not.

Reply:
continue takes effect only for the for loop, ref: https://github.com/golang/go/issues/43889.

Besides, only one select branch is executed in each for loop, so perhaps continue is not needed here?